lib/TH/THGeneral.c

```diff
@@ -138,20 +141,29 @@ static long getAllocSize(void *ptr) {
  * (2) if post-GC heap size exceeds 80% of the soft max, increase the
  *     soft max by 40%
  */
-static void maybeTriggerGC() {
-  if(torchGCFunction && torchHeapSize > torchHeapSizeSoftMax) {
+static void maybeTriggerGC(long curHeapSize) {
+  if(torchGCFunction && curHeapSize > heapSoftmax ) {
     torchGCFunction(torchGCData);
-    if(torchHeapSize > torchHeapSizeSoftMax * 0.8) {
-      torchHeapSizeSoftMax = torchHeapSizeSoftMax * 1.4;
+    long newHeapSize = THAtomicGetLong(&heapSize);
+    if(newHeapSize > heapSoftmax * heapSoftmaxGrowthThresh) {
+      heapSoftmax = heapSoftmax * heapSoftmaxGrowthFactor;
     }
   }
 }
 
 // hooks into the TH heap tracking
 void THHeapUpdate(long size) {
-  torchHeapSize += size;
-  if (size > 0)
-    maybeTriggerGC();
+  long newHeapSize = THAtomicAddLong(&heapSize, size);
```

Comment:
`THAtomicAddLong` returns the previous value of `heapSize`, so we have to add `size` to our local `newHeapSize`.